File: .claude/skills/lead-discovery/research-pipeline/company_enricher.py
#!/usr/bin/env python3
"""
Company Enricher - Enriches prospect data with web research
"""

import os
import json
import argparse
import asyncio
from typing import Optional
import aiohttp
from dotenv import load_dotenv

load_dotenv()

FIRECRAWL_API_KEY = os.getenv("FIRECRAWL_API_KEY")
PERPLEXITY_API_KEY = os.getenv("PERPLEXITY_API_KEY")


async def scrape_website(url: str) -> Optional[dict]:
    """Scrape company website using Firecrawl"""
    
    if not url or not FIRECRAWL_API_KEY:
        return None
    
    headers = {
        "Authorization": f"Bearer {FIRECRAWL_API_KEY}",
        "Content-Type": "application/json"
    }
    
    payload = {
        "url": url,
        "formats": ["markdown"],
        "onlyMainContent": True
    }
    
    async with aiohttp.ClientSession() as session:
        try:
            async with session.post(
                "https://api.firecrawl.dev/v1/scrape",
                headers=headers,
                json=payload,
                timeout=aiohttp.ClientTimeout(total=30)
            ) as response:
                if response.status == 200:
                    data = await response.json()
                    if data.get("success"):
                        return {
                            "content": data.get("data", {}).get("markdown", "")[:3000],
                            "title": data.get("data", {}).get("metadata", {}).get("title"),
                            "description": data.get("data", {}).get("metadata", {}).get("description")
                        }
        except Exception as e:
            print(f"Warning: Could not scrape {url}: {e}")
    
    return None


async def research_company(company_name: str, website: Optional[str]) -> Optional[dict]:
    """Research company using Perplexity"""
    
    if not PERPLEXITY_API_KEY:
        return None
    
    headers = {
        "Authorization": f"Bearer {PERPLEXITY_API_KEY}",
        "Content-Type": "application/json"
    }
    
    query = f"What does {company_name} do?"
    if website:
        query += f" (website: {website})"
    query += " Provide a brief description, their main products/services, and estimated company size."
    
    payload = {
        "model": "sonar",
        "messages": [
            {"role": "user", "content": query}
        ],
        "max_tokens": 300
    }
    
    async with aiohttp.ClientSession() as session:
        try:
            async with session.post(
                "https://api.perplexity.ai/chat/completions",
                headers=headers,
                json=payload,
                timeout=aiohttp.ClientTimeout(total=30)
            ) as response:
                if response.status == 200:
                    data = await response.json()
                    return {
                        "research": data.get("choices", [{}])[0].get("message", {}).get("content", "")
                    }
        except Exception as e:
            print(f"Warning: Could not research {company_name}: {e}")
    
    return None


async def enrich_prospect(prospect: dict) -> dict:
    """Enrich a single prospect with additional data"""
    
    enriched = prospect.copy()
    
    # Scrape website if available
    if prospect.get("website"):
        website_data = await scrape_website(prospect["website"])
        if website_data:
            enriched["website_content"] = website_data.get("content")
            enriched["website_title"] = website_data.get("title")
            enriched["website_description"] = website_data.get("description")
    
    # Research company
    research = await research_company(
        prospect.get("company_name", ""),
        prospect.get("website")
    )
    if research:
        enriched["research_summary"] = research.get("research")
    
    enriched["enriched"] = True
    
    return enriched


async def enrich_prospects(
    prospects: list,
    max_enrich: int = 20,
    delay: float = 1.0
) -> list:
    """Enrich multiple prospects"""
    
    enriched = []
    
    for i, prospect in enumerate(prospects):
        if i < max_enrich:
            print(f"Enriching {i+1}/{min(len(prospects), max_enrich)}: {prospect.get('company_name')}")
            enriched_prospect = await enrich_prospect(prospect)
            enriched.append(enriched_prospect)
            await asyncio.sleep(delay)  # Rate limiting
        else:
            # Keep remaining prospects without enrichment
            prospect["enriched"] = False
            enriched.append(prospect)
    
    return enriched


async def main():
    parser = argparse.ArgumentParser(description="Enrich prospect data")
    parser.add_argument("--input", required=True, help="Input prospects JSON file")
    parser.add_argument("--output", required=True, help="Output enriched JSON file")
    parser.add_argument("--max-enrich", type=int, default=20, help="Max prospects to enrich")
    
    args = parser.parse_args()
    
    try:
        # Load prospects
        with open(args.input, "r") as f:
            data = json.load(f)
        
        prospects = data.get("prospects", [])
        
        print(f"Enriching up to {args.max_enrich} of {len(prospects)} prospects...")
        
        # Enrich prospects
        enriched = await enrich_prospects(prospects, max_enrich=args.max_enrich)
        
        # Prepare output
        output = {
            "query": data.get("query", {}),
            "total_prospects": len(enriched),
            "enriched_count": sum(1 for p in enriched if p.get("enriched")),
            "prospects": enriched
        }
        
        # Ensure output directory exists
        os.makedirs(os.path.dirname(args.output), exist_ok=True)
        
        # Save results
        with open(args.output, "w") as f:
            json.dump(output, f, indent=2)
        
        print(f"Enriched results saved to: {args.output}")
        return 0
        
    except Exception as e:
        print(f"Error: {e}")
        return 1


if __name__ == "__main__":
    exit(asyncio.run(main()))

File: packages/agent/src/tools/company-enricher.ts
/**
 * Company Enricher Tool - Enriches company data using Firecrawl and Perplexity
 */

import { tool } from 'ai';
import { z } from 'zod';
import { normalizeWebsiteUrl } from '../utils/url.js';

const FIRECRAWL_API_URL = 'https://api.firecrawl.dev/v1';
const PERPLEXITY_API_URL = 'https://api.perplexity.ai/chat/completions';

// Configuration
const DEFAULT_CONCURRENCY = 3;

// Structured logging helper
function log(event: string, data: Record<string, unknown>) {
  console.log(JSON.stringify({
    timestamp: new Date().toISOString(),
    component: 'CompanyEnricher',
    event,
    ...data,
  }));
}

/**
 * Simple semaphore for concurrency limiting
 */
function createSemaphore(maxConcurrency: number) {
  let running = 0;
  const queue: Array<() => void> = [];

  return {
    async acquire(): Promise<void> {
      if (running < maxConcurrency) {
        running++;
        return;
      }
      return new Promise(resolve => queue.push(resolve));
    },
    release(): void {
      running--;
      const next = queue.shift();
      if (next) {
        running++;
        next();
      }
    },
  };
}

interface Prospect {
  id: string;
  company_name: string;
  domain?: string;
  website?: string;
  [key: string]: unknown;
}

interface EnrichedProspect extends Prospect {
  website_content?: string;
  company_summary?: string;
  enrichment_status: 'success' | 'partial' | 'failed';
}

async function scrapeWebsite(website: string, apiKey: string, companyName: string): Promise<string | null> {
  const cleanUrl = normalizeWebsiteUrl(website);
  const startTime = Date.now();

  log('firecrawl_start', { company: companyName, url: cleanUrl });

  try {
    const response = await fetch(`${FIRECRAWL_API_URL}/scrape`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${apiKey}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        url: cleanUrl,
        formats: ['markdown'],
        onlyMainContent: true,
      }),
      signal: AbortSignal.timeout(30000), // 30s timeout
    });

    if (!response.ok) {
      log('firecrawl_error', {
        company: companyName,
        url: cleanUrl,
        status: response.status,
        durationMs: Date.now() - startTime,
      });
      return null;
    }

    const data = await response.json() as { data?: { markdown?: string } };
    const content = data.data?.markdown?.substring(0, 2000) || null;

    log('firecrawl_complete', {
      company: companyName,
      url: cleanUrl,
      durationMs: Date.now() - startTime,
      contentLength: content?.length || 0,
      success: !!content,
    });

    return content;
  } catch (error) {
    log('firecrawl_error', {
      company: companyName,
      url: cleanUrl,
      error: error instanceof Error ? error.message : String(error),
      durationMs: Date.now() - startTime,
    });
    return null;
  }
}

async function researchCompany(companyName: string, apiKey: string): Promise<string | null> {
  const startTime = Date.now();

  log('perplexity_start', { company: companyName });

  try {
    const response = await fetch(PERPLEXITY_API_URL, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${apiKey}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        model: 'sonar',
        messages: [
          {
            role: 'system',
            content: 'You are a business analyst. Provide a brief 2-3 sentence summary of the company.',
          },
          {
            role: 'user',
            content: `Provide a brief summary of ${companyName}. What do they do and who are their customers?`,
          },
        ],
        max_tokens: 200,
      }),
      signal: AbortSignal.timeout(30000), // 30s timeout
    });

    if (!response.ok) {
      log('perplexity_error', {
        company: companyName,
        status: response.status,
        durationMs: Date.now() - startTime,
      });
      return null;
    }

    const data = await response.json() as { choices?: Array<{ message?: { content?: string } }> };
    const summary = data.choices?.[0]?.message?.content || null;

    log('perplexity_complete', {
      company: companyName,
      durationMs: Date.now() - startTime,
      summaryLength: summary?.length || 0,
      success: !!summary,
    });

    return summary;
  } catch (error) {
    log('perplexity_error', {
      company: companyName,
      error: error instanceof Error ? error.message : String(error),
      durationMs: Date.now() - startTime,
    });
    return null;
  }
}

async function enrichProspect(
  prospect: Prospect,
  firecrawlKey: string | undefined,
  perplexityKey: string | undefined
): Promise<EnrichedProspect> {
  const startTime = Date.now();

  log('enrich_start', {
    company: prospect.company_name,
    website: prospect.website,
    hasFirecrawl: !!firecrawlKey,
    hasPerplexity: !!perplexityKey,
  });

  let websiteContent: string | null = null;
  let companySummary: string | null = null;

  // Run Firecrawl and Perplexity in parallel for better performance
  const [firecrawlResult, perplexityResult] = await Promise.all([
    prospect.website && firecrawlKey
      ? scrapeWebsite(prospect.website, firecrawlKey, prospect.company_name)
      : Promise.resolve(null),
    perplexityKey
      ? researchCompany(prospect.company_name, perplexityKey)
      : Promise.resolve(null),
  ]);

  websiteContent = firecrawlResult;
  companySummary = perplexityResult;

  const hasWebsite = !!websiteContent;
  const hasSummary = !!companySummary;
  const status = hasWebsite && hasSummary ? 'success' : (hasWebsite || hasSummary ? 'partial' : 'failed');

  log('enrich_complete', {
    company: prospect.company_name,
    totalDurationMs: Date.now() - startTime,
    status,
    hasWebsiteContent: hasWebsite,
    hasSummary,
    websiteContentLength: websiteContent?.length || 0,
  });

  return {
    ...prospect,
    website_content: websiteContent || undefined,
    company_summary: companySummary || undefined,
    enrichment_status: status,
  };
}

// Standalone execute function for direct use in pipelines
export async function executeCompanyEnricher({
  prospects,
  maxProcess = 20,
  maxConcurrency = DEFAULT_CONCURRENCY,
  delayMs = 500,
}: {
  prospects: Array<{
    id: string;
    company_name: string;
    domain?: string;
    website?: string;
    [key: string]: unknown;
  }>;
  maxProcess?: number;
  maxConcurrency?: number;
  delayMs?: number;
}) {
  const firecrawlKey = process.env.FIRECRAWL_API_KEY;
  const perplexityKey = process.env.PERPLEXITY_API_KEY;

  if (!firecrawlKey && !perplexityKey) {
    return {
      success: false as const,
      error: 'Neither FIRECRAWL_API_KEY nor PERPLEXITY_API_KEY configured',
    };
  }

  const toProcess = prospects.slice(0, maxProcess);
  const remaining = prospects.slice(maxProcess);

  log('batch_start', {
    totalProspects: prospects.length,
    toProcess: toProcess.length,
    maxConcurrency,
    delayMs,
  });

  const batchStartTime = Date.now();
  const semaphore = createSemaphore(maxConcurrency);

  // Process prospects in parallel with concurrency limit
  const enrichmentPromises = toProcess.map(async (prospect, index) => {
    await semaphore.acquire();

    try {
      // Stagger start times slightly to avoid API burst
      if (index > 0) {
        await new Promise(resolve => setTimeout(resolve, delayMs * (index % maxConcurrency)));
      }

      return await enrichProspect(
        prospect as Prospect,
        firecrawlKey,
        perplexityKey
      );
    } finally {
      semaphore.release();
    }
  });

  const enriched = await Promise.all(enrichmentPromises);

  // Add remaining prospects without enrichment
  for (const prospect of remaining) {
    enriched.push({
      ...(prospect as Prospect),
      enrichment_status: 'failed',
    });
  }

  const successCount = enriched.filter(p => p.enrichment_status === 'success').length;
  const partialCount = enriched.filter(p => p.enrichment_status === 'partial').length;
  const failedCount = enriched.filter(p => p.enrichment_status === 'failed').length;

  log('batch_complete', {
    totalDurationMs: Date.now() - batchStartTime,
    total: enriched.length,
    successful: successCount,
    partial: partialCount,
    failed: failedCount,
    avgTimePerProspect: Math.round((Date.now() - batchStartTime) / toProcess.length),
  });

  return {
    success: true as const,
    total_prospects: enriched.length,
    enriched_count: successCount + partialCount,
    full_enrichment: successCount,
    partial_enrichment: partialCount,
    prospects: enriched,
  };
}

export const companyEnricherTool = tool({
  description: 'Enrich prospect data with website content and company research from Firecrawl and Perplexity',
  inputSchema: z.object({
    prospects: z.array(z.object({
      id: z.string(),
      company_name: z.string(),
      domain: z.string().optional(),
      website: z.string().optional(),
    }).passthrough()).describe('Array of prospects to enrich'),
    maxProcess: z.number().default(20).describe('Maximum number of prospects to process'),
    delayMs: z.number().default(1000).describe('Delay between API calls in milliseconds'),
  }),
  execute: async ({ prospects, maxProcess, delayMs }) => {
    const firecrawlKey = process.env.FIRECRAWL_API_KEY;
    const perplexityKey = process.env.PERPLEXITY_API_KEY;

    if (!firecrawlKey && !perplexityKey) {
      return {
        success: false,
        error: 'Neither FIRECRAWL_API_KEY nor PERPLEXITY_API_KEY configured',
      };
    }

    const enriched: EnrichedProspect[] = [];
    const toProcess = prospects.slice(0, maxProcess);
    const remaining = prospects.slice(maxProcess);

    for (const prospect of toProcess) {
      const result = await enrichProspect(
        prospect as Prospect,
        firecrawlKey,
        perplexityKey
      );
      enriched.push(result);

      // Rate limiting
      await new Promise(resolve => setTimeout(resolve, delayMs));
    }

    // Add remaining prospects without enrichment
    for (const prospect of remaining) {
      enriched.push({
        ...(prospect as Prospect),
        enrichment_status: 'failed',
      });
    }

    const successCount = enriched.filter(p => p.enrichment_status === 'success').length;
    const partialCount = enriched.filter(p => p.enrichment_status === 'partial').length;

    return {
      success: true,
      total_prospects: enriched.length,
      enriched_count: successCount + partialCount,
      full_enrichment: successCount,
      partial_enrichment: partialCount,
      prospects: enriched,
    };
  },
});
